examples/vmwithcontracts/actions/call.go

```diff
@@ -0,0 +1,133 @@
+// Copyright (C) 2023, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package actions
+
+import (
+	"context"
+
+	"github.com/ava-labs/avalanchego/ids"
+	"github.com/ava-labs/avalanchego/utils/units"
+
+	"github.com/ava-labs/hypersdk/chain"
+	"github.com/ava-labs/hypersdk/codec"
+	"github.com/ava-labs/hypersdk/consts"
+	"github.com/ava-labs/hypersdk/examples/vmwithcontracts/storage"
+	"github.com/ava-labs/hypersdk/state"
+	"github.com/ava-labs/hypersdk/x/programs/runtime"
+
+	mconsts "github.com/ava-labs/hypersdk/examples/vmwithcontracts/consts"
+)
+
+var _ chain.Action = (*Call)(nil)
+
+type StateKeyPermission struct {
+	Key        string
+	Permission state.Permissions
+}
+
+type Call struct {
+	// program is the address of the program to be called
+	ContractAddress codec.Address `json:"contractAddress"`
+
+	// Amount are transferred to [To].
+	Value uint64 `json:"value"`
+
+	// Function is the name of the function to call on the program.
+	Function string `json:"function"`
+
+	// CallData are the serialized parameters to be passed to the program.
+	CallData []byte `json:"calldata"`
+
+	SpecifiedStateKeys []StateKeyPermission `json:"statekeys"`
+
+	Fuel uint64 `json:"fuel"`
+}
+
+func (*Call) GetTypeID() uint8 {
+	return mconsts.CallProgramID
+}
+
+func (t *Call) StateKeys(_ codec.Address, _ ids.ID) state.Keys {
+	result := state.Keys{}
+	for _, stateKeyPermission := range t.SpecifiedStateKeys {
+		result.Add(stateKeyPermission.Key, stateKeyPermission.Permission)
+	}
+	return result
+}
```

Comment:
Note: this only uses the last permissions value specified, which may produce unexpected behavior

Reply:
Not sure what you mean by this

Reply:
if `Call` specifies 2 instances of the same state key with a different value for the `Permissions` field then during this iteration the first will be overwritten by the second. If we ever changed this in the future, it would be a breaking change.

Reply:
ahh, sure.  This is usually going to be built from a map, but you could have duplicate keys.  It wouldn't overwrite, it would union.  Not sure that is any more expected, but it seems reasonable to me.

Reply:
@aaronbuchwald, I can create an issue to fix this (ie check for uniqueness) after this PR is merged.